x/mongo/driver/session/client_session.go

```diff
@@ -215,16 +215,31 @@ func NewClientSession(pool *Pool, clientID uuid.UUID, sessionType Type, opts ...
 		return nil, errors.New("causal consistency and snapshot cannot both be set for a session")
 	}
 
-	servSess, err := pool.GetSession()
-	if err != nil {
-		return nil, err
+	if sessionType != Implicit || !mergedOpts.deferImplicit {
```

Comment:
```suggestion
	if sessionType != Implicit {
```
I'm a little confused why we need the new `deferImplicit` field on `session.ClientOptions`. Shouldn't we just only `SetServer` if it's an explicit session here, and otherwise wait until `operation.Execute` to call `SetServer` (after connection checkout).

Reply:
I think we want to default the behavior of `NewClientSession` to create an implicit session, as before. This ensures that existing functionality is preserved. The additional non-exported option allows us to extend the logic, however, wrapping this non-default case as the default behavior for `NewImplicitClientSessionDeferred`.

Reply:
Okay, sounds good.

Reply:
@benjirewis @matthewdale This change causes some failures in legacy write concern tests: https://evergreen.mongodb.com/build/mongo_go_driver_tests_42_plus_zlib_zstd_support__version~6.0_os_ssl_40~ubuntu1804_64_go_1_17_patch_f87317a97462447361004797b2e6d1023c61eaf4_62e008d13066152bc52df78e_22_07_26_15_31_48

Will have to investigate exactly why, but I'm sure there are collection methods, such as drop and aggregate, that can't defer implicit sessions in this exact way.

Reply:
This issue should be resolved by [this](https://github.com/mongodb/mongo-go-driver/pull/1027/files#diff-167dc8480d4039f98a2750ef31b20aac979efbb1c83466ac18cbf12c96b72b94R526-R533) block.